lib/tty/pager/basic.rb

```diff
@@ -43,48 +45,112 @@ def default_prompt
       #
       # @api public
       def page(text, &callback)
-        page_num = 1
-        leftover = []
-        lines_left = @height
+        write(text, &callback)
+      rescue PagerClosed
+        # do nothing
+      ensure
+        reset
+      end
+
+      # Write text to the pager, prompting on page end.
+      #
+      # @raise [PagerClosed]
+      #   if the pager was closed
+      #
+      # @return [TTY::Pager::BasicPager]
+      #
+      # @api public
+      def write(*args, &callback)
+        args.each do |text|
+          send_text(:write, text, &callback)
+        end
+        self
+      end
+      alias_method :<<, :write
+
+      # Write text to the pager, prompting on page end. Returns false if the
+      # pager was closed.
+      #
+      # @return [Boolean]
+      #   the success status of writing to the screen
+      #
+      # @api public
+      def try_write(*args, &callback)
+        write(text, &callback)
+        true
+      rescue PagerClosed
+        false
+      end
 
+      # Print a line of text to the pager, prompting on page end.
+      #
+      # @raise [PagerClosed]
+      #   if the pager was closed
+      #
+      # @api public
+      def puts(text, &callback)
+        send_text(:puts, text, &callback)
+      end
+
+      # The lower-level common implementation of printing methods
+      #
+      # @return [Boolean]
+      #   the success status of writing to the screen
+      #
+      # @api private
+      def send_text(write_method, text, &callback)
         text.lines.each do |line|
           chunk = []
-          if !leftover.empty?
-            chunk = leftover
-            leftover = []
+          if !@leftover.empty?
+            chunk = @leftover
+            @leftover = []
           end
           wrapped_line = Strings.wrap(line, @width)
           wrapped_line.lines.each do |line_part|
-            if lines_left > 0
+            if @lines_left > 0
               chunk << line_part
-              lines_left -= 1
+              @lines_left -= 1
             else
-              leftover << line_part
+              @leftover << line_part
             end
           end
-          output.print(chunk.join)
+          output.public_send(write_method, chunk.join)
 
-          if lines_left == 0
-            break unless continue_paging?(page_num)
-            lines_left = @height
-            if leftover.size > 0
-              lines_left -= leftover.size
+          if @lines_left == 0
+            raise PagerClosed unless continue_paging?(@page_num)
+            @lines_left = @height
+            if @leftover.size > 0
+              @lines_left -= @leftover.size
             end
-            page_num += 1
-            return !callback.call(page_num) unless callback.nil?
+            @page_num += 1
+            callback.call(@page_num) unless callback.nil?
```

Comment:
I'm not sure what the idea behind this callback is. I guess there should be a `return` here to exit the function, but I'm not sure about the conditions. What's the case for invoking `write` with a callback?